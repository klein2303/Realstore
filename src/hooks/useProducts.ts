import {useQuery, UseQueryResult} from "@tanstack/react-query";
import {Product} from "../models/Product";

const fetchProducts = async (
    category?: string,
    productId?: number,
    minPrice?: number,
    maxPrice?: number,
    minRating?: number,
): Promise<Product[]> => {
    let url = "https://fakestoreapi.com/products";

    if (productId) {
        url = `${url}/${productId}`;
        const response = await fetch(url);
        if (!response.ok) throw new Error("Error fetching product");
        const product: Product = await response.json();
        return [product]; // Return a single product wrapped in an array
    } else if (category) {
        url = `${url}/category/${category}`;
    }

    const response = await fetch(url);
    if (!response.ok) throw new Error("Error fetching products");
    let products: Product[] = await response.json();

    // Filter products based on price and rating
    if (minPrice !== undefined) {
        products = products.filter((product) => product.price >= minPrice);
    }
    if (maxPrice !== undefined) {
        products = products.filter((product) => product.price <= maxPrice);
    }
    if (minRating !== undefined) {
        products = products.filter((product) => product.rating.rate >= minRating);
    }

    return products;
};

// Custom hook to fetch products with optional filters
export const useProducts = (
    category?: string,
    productId?: number,
    minPrice?: number,
    maxPrice?: number,
    minRating?: number,
): UseQueryResult<Product[]> => {
    const queryKey = ["products", category, productId, minPrice, maxPrice, minRating];

    return useQuery<Product[]>({
        queryKey,
        queryFn: () => fetchProducts(category, productId, minPrice, maxPrice, minRating),
    });
};
