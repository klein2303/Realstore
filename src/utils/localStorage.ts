// Function to store favorite items in local storage
export const setFavorites = (items: number[]): void => {
    localStorage.setItem("favorites", JSON.stringify(items));
};

// Function to get favorite items from local storage
export const getFavorites = (): number[] => {
    const storedItems = localStorage.getItem("favorites");
    return storedItems ? JSON.parse(storedItems) : [];
};

export const toggleFavorite = (productId: number): void => {
    const favorites = getFavorites();
    const isFavorite = favorites.includes(productId);

    const updatedFavorites = isFavorite ? favorites.filter((id) => id !== productId) : [...favorites, productId];

    setFavorites(updatedFavorites);
};
