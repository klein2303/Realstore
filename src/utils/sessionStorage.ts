import {Filter} from "../pages/Shopping";

export type CartItem = {productId: number; quantity: number};

// Helper function to store items in session storage
const setSessionStorageItems = <T>(key: string, items: T[]): void => {
    sessionStorage.setItem(key, JSON.stringify(items));
};

// Helper function to get items from session storage
const getSessionStorageItems = <T>(key: string): T[] => {
    const storedItems = sessionStorage.getItem(key);
    return storedItems ? JSON.parse(storedItems) : [];
};

// Cart-specific functions
export const getCart = (): CartItem[] => {
    return getSessionStorageItems<CartItem>("cart");
};

export const addToCart = (productId: number, quantity: number = 1): void => {
    const cart = getCart();
    const existingItem = cart.find((item) => item.productId === productId);

    if (existingItem) {
        existingItem.quantity += quantity; // Increment quantity
    } else {
        cart.push({productId, quantity});
    }

    setSessionStorageItems("cart", cart);
};

export const clearCart = (): void => {
    setSessionStorageItems("cart", []);
};

export const updateCartQuantity = (productId: number, quantity: number): void => {
    const cart = getCart();
    const index = cart.findIndex((item) => item.productId === productId);

    if (index !== -1) {
        // Product exists in the cart
        if (quantity > 0) {
            cart[index].quantity = quantity; // Update quantity
        } else {
            cart.splice(index, 1); // Remove if quantity is 0
        }
    } else if (quantity > 0) {
        // Product doesn't exist in the cart, add it
        cart.push({productId, quantity});
    }

    setSessionStorageItems("cart", cart); // Save updated cart to sessionStorage
};

// Filtered items-specific functions
export const getFilter = (): Filter[] => {
    return getSessionStorageItems<Filter>("filter");
};

export const setFilter = (filter: Filter[]): void => {
    setSessionStorageItems("filter", filter);
};

export const clearFilter = (): void => {
    setSessionStorageItems("filter", []);
};
