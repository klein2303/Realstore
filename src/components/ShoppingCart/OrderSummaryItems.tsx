import styles from "./OrderSummaryItems.module.css";
import {Product} from "../../models/Product";
import {getCart, CartItem} from "../../utils/sessionStorage";
import {useProducts} from "../../hooks/useProducts.ts";
import {useEffect, useState} from "react";
import ClothingsCards from "../Scrolling/ClothingsCards/ClothingsCards";

const OrderSummaryItems = () => {
    const [cartItemData, setCartItemData] = useState<CartItem[]>([]);
    const [cartItems, setCartItems] = useState<Product[]>([]);
    const {data: products, isLoading, isError} = useProducts();

    useEffect(() => {
        if (!products) {
            return;
        }
        const cartItemData = getCart();
        setCartItemData(cartItemData);
        const cartItemIds = cartItemData.map((cartId: CartItem) => {
            return cartId.productId;
        });
        const cartItems = products.filter((product: Product) => cartItemIds.includes(product.id));
        setCartItems(cartItems);
    }, [products]);

    if (isLoading) {
        return (
            <section role="status" aria-live="polite" aria-label="Loading">
                Loading...
            </section>
        );
    }
    if (isError) {
        return (
            <section role="alert" aria-live="assertive" aria-label="Error fetching products">
                Error fetching products.
            </section>
        );
    }

    return (
        <>
            <section className={styles.itembox} role="region" aria-label="Order Summary Items">
                {cartItemData.length === 0 ? (
                    <section
                        className={styles.message}
                        role="status"
                        aria-live="polite"
                        aria-label="No products available">
                        No products available.
                    </section>
                ) : (
                    cartItems.map((product: Product) => (
                        <ClothingsCards
                            key={product.id}
                            id={product.id}
                            title={product.title.substring(0, 20) + "..."}
                            price={product.price}
                            category={product.category.charAt(0).toUpperCase() + product.category.substring(1)}
                            image={product.image}
                            cart={true}
                            favorite={false}
                        />
                    ))
                )}
            </section>
        </>
    );
};

export default OrderSummaryItems;
