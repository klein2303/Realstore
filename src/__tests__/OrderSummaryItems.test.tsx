import {render, screen} from "@testing-library/react";
import '@testing-library/jest-dom';
import {describe, it, expect, vi, beforeEach} from "vitest";
import OrderSummaryItems from "../components/ShoppingCart/OrderSummaryItems";
import {useProducts} from "../hooks/useProducts";
import {getCart} from "../utils/sessionStorage";
import {MemoryRouter} from "react-router-dom";

vi.mock("../hooks/useProducts", () => ({
    useProducts: vi.fn(),
}));

vi.mock("../utils/sessionStorage", () => ({
    getCart: vi.fn(),
}));

describe("OrderSummaryItems Component", () => {
    beforeEach(() => {
        vi.resetAllMocks();
    });

    it("renders the loading state", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: null,
            isLoading: true,
            isError: false,
        });

        render(
            <MemoryRouter>
                <OrderSummaryItems />
            </MemoryRouter>,
        );

        expect(screen.getByRole("status", {name: /loading/i})).toBeInTheDocument();
    });

    it("renders the error state", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: null,
            isLoading: false,
            isError: true,
        });

        render(
            <MemoryRouter>
                <OrderSummaryItems />
            </MemoryRouter>,
        );

        expect(screen.getByRole("alert", {name: /error fetching products/i})).toBeInTheDocument();
    });

    it("renders 'No products available' when cart is empty", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: [],
            isLoading: false,
            isError: false,
        });

        (getCart as jest.Mock).mockReturnValue([]);

        render(
            <MemoryRouter>
                <OrderSummaryItems />
            </MemoryRouter>,
        );

        expect(screen.getByRole("status", {name: /no products available/i})).toBeInTheDocument();
    });

    it("renders products from the cart", () => {
        const mockProducts = [
            {
                id: 1,
                title: "Mock Product 1",
                price: 100,
                category: "electronics",
                image: "mock-image-1.jpg",
            },
            {
                id: 2,
                title: "Mock Product 2",
                price: 200,
                category: "clothing",
                image: "mock-image-2.jpg",
            },
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: mockProducts,
            isLoading: false,
            isError: false,
        });

        (getCart as jest.Mock).mockReturnValue([{productId: 1}, {productId: 2}]);

        render(
            <MemoryRouter>
                <OrderSummaryItems />
            </MemoryRouter>,
        );

        expect(screen.getByRole("region", {name: /order summary items/i})).toBeInTheDocument();
        expect(screen.getAllByRole("region", {name: /product card for/i})).toHaveLength(2);
    });

    it("matches the snapshot when products are available", () => {
        const mockProducts = [
            {
                id: 1,
                title: "Mock Product 1",
                price: 100,
                category: "electronics",
                image: "mock-image-1.jpg",
            },
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: mockProducts,
            isLoading: false,
            isError: false,
        });

        (getCart as jest.Mock).mockReturnValue([{productId: 1}]);

        const {container} = render(
            <MemoryRouter>
                <OrderSummaryItems />
            </MemoryRouter>,
        );

        expect(container).toMatchSnapshot();
    });
});
