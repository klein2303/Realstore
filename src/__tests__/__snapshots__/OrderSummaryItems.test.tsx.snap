// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OrderSummaryItems Component > matches the snapshot when products are available 1`] = `
<div>
  <section
    aria-label="Order Summary Items"
    class="_itembox_5c914e"
    role="region"
  >
    <article
      aria-label="Product card for Mock Product 1..."
      class="_clothingSection_47ab48"
      role="region"
    >
      <figure
        class="_imgContainer_47ab48"
      >
        <a
          href="/Clothe/1"
        >
          <img
            alt="Image of Mock Product 1..."
            class="_clothingImage_47ab48"
            src="mock-image-1.jpg"
          />
        </a>
      </figure>
      <section
        class="_descriptionSpace_47ab48"
      >
        <header
          class="_productType_47ab48"
        >
          <h5
            class="_category_47ab48"
          >
            Electronics
          </h5>
          <section
            aria-label="Product Actions"
            class="_IconSpace_47ab48"
          >
            <span
              aria-label="Remove from cart"
              role="button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                tabindex="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="176"
                  cy="416"
                  r="32"
                />
                <circle
                  cx="400"
                  cy="416"
                  r="32"
                />
                <path
                  d="M456.8 120.78a23.92 23.92 0 0 0-18.56-8.78H133.89l-6.13-34.78A16 16 0 0 0 112 64H48a16 16 0 0 0 0 32h50.58l45.66 258.78A16 16 0 0 0 160 368h256a16 16 0 0 0 0-32H173.42l-5.64-32h241.66A24.07 24.07 0 0 0 433 284.71l28.8-144a24 24 0 0 0-5-19.93z"
                />
              </svg>
            </span>
          </section>
        </header>
        <footer
          class="_nameAndPrice_47ab48"
        >
          <span
            class="_productname_47ab48"
          >
            <h4
              class="_h4text_47ab48"
            >
              Mock Product 1...
            </h4>
          </span>
          <span
            class="_price_47ab48"
          >
            <h4
              class="_h4text_47ab48"
            >
              $ 
              100
            </h4>
          </span>
        </footer>
      </section>
    </article>
  </section>
</div>
`;
